crates/ruff_python_formatter/src/expression/mod.rs

```diff
@@ -576,23 +582,66 @@ impl CallChainLayout {
     }
 }
 
-fn has_parentheses(expr: &Expr, source: &str) -> bool {
-    has_own_parentheses(expr) || is_expression_parenthesized(AnyNodeRef::from(expr), source)
+#[derive(Debug, Copy, Clone, PartialEq, Eq, is_macro::Is)]
+pub(crate) enum OwnParentheses {
+    /// The node has parentheses, but they are empty (e.g., `[]` or `f()`).
+    Empty,
+    /// The node has parentheses, and they are non-empty (e.g., `[1]` or `f(1)`).
+    NonEmpty,
 }
 
-pub(crate) const fn has_own_parentheses(expr: &Expr) -> bool {
-    matches!(
-        expr,
-        Expr::Dict(_)
-            | Expr::List(_)
-            | Expr::Tuple(_)
-            | Expr::Set(_)
-            | Expr::ListComp(_)
-            | Expr::SetComp(_)
-            | Expr::DictComp(_)
-            | Expr::Call(_)
-            | Expr::Subscript(_)
-    )
+fn has_parentheses(expr: &Expr, source: &str) -> Option<OwnParentheses> {
+    if is_expression_parenthesized(AnyNodeRef::from(expr), source) {
+        Some(OwnParentheses::NonEmpty)
+    } else {
+        has_own_parentheses(expr, source)
+    }
```

Comment:
i'd switch the checks because i expect `has_own_parentheses` to be faster and be `Some(_)` more often

Reply:
Yeah it's just a little tricky because if `has_own_parentheses` tells us the node has parentheses, but they're empty, then we still need to check `is_expression_parenthesized` to handle cases like `([])` (needs to return non-empty, even though the node's _own_ parentheses are empty). I reversed the cases, it's just more branches.